extensions/applicationinsights-debugplugin-js/src/DebugPlugin.ts

```diff
@@ -0,0 +1,295 @@
+          const defaultConfig = DebugPlugin.getDefaultConfig();
+          const ctx = _self._getTelCtx();
+          const identifier = _self.identifier;
+          for (const field in defaultConfig) {
+              _extensionConfig[field] = () => ctx.getConfig(identifier, field, defaultConfig[field]());
+          }
+
+          trackers = _extensionConfig.trackers();
+
+          // 3. setup debugBin
+          const debugBinContainer = document.createElement("div");
+          debugBinContainer.className = 'debug-bin-container';
+          debugBinParent = new DebugBinParent(debugBinContainer, [], 0);
+
+          // 4. grab all relevant functions that need to be instrumented and store them in arrays
+          const diagLog = _self.diagLog();
+
+          debugBins = {};
+          const propertiesProtoFns: string[] = [];
+          const analyticsProtoFns: string[] = [];
+          const ajaxProtoFns: string[] = [];
+          const channelProtoFns: string[] = [];
+          const diagLogProtoFns: string[] = [];
+          for (const [ext, protoFns] of [
+            [analyticsExt, analyticsProtoFns],
+            [propertiesExt, propertiesProtoFns],
+            [ajaxDependencyExt, ajaxProtoFns],
+            [channelPluginExt, channelProtoFns],
+            [diagLog, diagLogProtoFns]
+          ] as any[]) {
+            for (const key of CoreUtils.objKeys(ext['__proto__'])) {
+              if (key.substring(0, 1) === '_') { continue; }
+              if (CoreUtils.isTypeof(ext[key], 'function')) {
+                protoFns.push(key);
+              }
+            }
+            // special case for sender
+            if (ext.identifier === 'AppInsightsChannelPlugin' && CoreUtils.arrIndexOf(trackers, '_sender') !== -1) {
+              protoFns.push('_sender');
+            }
+          }
+
+          // 5. actually instrument all the functions
+          for (let i = 0; i < trackers.length; i++) {
+            const tracker = trackers[i];
+            let target;
+            if (CoreUtils.arrIndexOf(propertiesProtoFns, tracker) !== -1) { target = propertiesExt['__proto__'] }
+            else if (CoreUtils.arrIndexOf(analyticsProtoFns, tracker) !== -1) { target = analyticsExt['__proto__'] }
+            else if (CoreUtils.arrIndexOf(ajaxProtoFns, tracker) !== -1) { target = ajaxDependencyExt['__proto__'] }
+            else if (CoreUtils.arrIndexOf(diagLogProtoFns, tracker) !== -1) { target = diagLog['__proto__'] }
+            // special case for sender
+            else if (tracker === '_sender') { target = channelPluginExt }
+            else if (CoreUtils.arrIndexOf(channelProtoFns, tracker) !== -1) { target = channelPluginExt['__proto__'] }
+            else { continue; }
+            InstrumentFunc(target, tracker, {
+              req: _self.preProcessItem(tracker) as any as () => InstrumentorHooksCallback,
+              rsp: _self.postProcessItem(tracker) as any as () => InstrumentorHooksCallback
+            });
```

Comment:
This seems overly complex, as I though I built some of these protections into the InstrumentFunc, but we can review later.